### PR TITLE
[Feature] Support dynamic partition pruning

### PR DESCRIPTION
**Pull request by letian-jiang**
Open-sourced table formats (e.g. hive) support data partitioning. That is to say, any record in a given data file has the same set of partition values.

This is a common practice for large fact tables (e.g. `lineorder` can be partitioned by `lo_orderdate` in the SSB dataset ). 

If the predicates generated by the runtime filter (`lo_orderdate IN SET(...)`) can be applied to the partition columns, we could use this to test if a data file (i.e. scan range) can be skipped or not.

SSB Q1
image

image

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 8 / 31 (25.81%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/connector/hive_connector.cpp|8|31|25.81%|[205, 208, 211, 214, 223, 224, 225, 228, 229, 231, 232, 233, 234, 235, 236, 241, 242, 243, 244, 245, 248, 249, 250]|

File: be/src/connector/hive_connector.cpp [hunk 1]
```diff
@@ -162,44 +165,46 @@ Status HiveDataSource::_init_partition_values() {
...
-            const ColumnPtr& data_column = const_column->data_column();
-            ColumnPtr& chunk_part_column = partition_chunk->get_column_by_slot_id(slot_id);
-            if (data_column->is_nullable()) {
-                chunk_part_column->append_default();
-            } else {
-                chunk_part_column->append(*data_column, 0, 1);
+    _partition_values = partition_desc->partition_key_value_evals();
+
+    // init partition chunk
+    auto partition_chunk = std::make_shared<Chunk>();
```
Comment (dirtysalt) [inline be/src/connector/hive_connector.cpp:171]: you'd better to stick to old implementation.

File: be/src/connector/hive_connector.cpp [hunk 2]
```diff
@@ -162,44 +165,46 @@ Status HiveDataSource::_init_partition_values() {
...
-            RETURN_IF_ERROR(ExecNode::eval_conjuncts(ctxs, partition_chunk.get()));
-        } else {
-            RETURN_IF_ERROR(ExecNode::eval_conjuncts(_partition_conjunct_ctxs, partition_chunk.get()));
-        }
+    if (!partition_chunk->has_rows()) {
+        _filter_by_eval_partition_conjuncts = true;
+        return Status::OK();
+    }
 
+    if (_enable_dynamic_prune_scan_range && partition_chunk && _runtime_filters) {
```
Comment (dirtysalt) [inline be/src/connector/hive_connector.cpp:199]: `partition_chunk` is always not null right?

File: be/src/formats/parquet/file_reader.cpp [hunk 3]
```diff
@@ -209,22 +209,6 @@ StatusOr<bool> FileReader::_filter_group(const tparquet::RowGroup& row_group) {
             if (discard) {
                 return true;
             }
-
-            // skip topn runtime filter, because it has taken effect on min/max filtering
-            // if row-group contains exactly one value(i.e. min_value = max_value), use bloom filter to test
-            if (!filter->always_true() && min_chunk->columns()[0]->equals(0, *max_chunk->columns()[0], 0)) {
```
Comment (dirtysalt) [inline be/src/formats/parquet/file_reader.cpp:215]: why remove this?

File: be/src/connector/hive_connector.cpp (hunk 2 above)
Comment (letian-jiang) [reply be/src/connector/hive_connector.cpp:199]: Good catch

File: be/src/connector/hive_connector.cpp (hunk 1 above)
Comment (letian-jiang) [reply be/src/connector/hive_connector.cpp:171]: I think both implementation are equivalent.

File: be/src/formats/parquet/file_reader.cpp (hunk 3 above)
Comment (letian-jiang) [reply be/src/formats/parquet/file_reader.cpp:215]: This is done for pruning iceberg table with partition column. Can be superseded by DPP

File: be/src/exprs/runtime_filter_bank.cpp [hunk 4]
```diff
@@ -417,6 +417,64 @@ void RuntimeFilterProbeCollector::do_evaluate(Chunk* chunk, RuntimeBloomFilterEv
...
+        }
+
+        auto* probe_expr = rf_desc->probe_expr_ctx();
+        auto* partition_by_exprs = rf_desc->partition_by_expr_contexts();
+
+        // skip if (conditions can be relaxed later)
+        //  1. multi-column runtime filter
+        //  2. runtime filter that is not slot ref, or a slot ref of a non-existent column
+        auto can_use_rf_on_partial_chunk = [&]() {
+            if (!partition_by_exprs->empty()) {
```
Comment (satanson) [inline be/src/exprs/runtime_filter_bank.cpp:444]:
```cpp
if (!partition_by_exprs->empty()) {
    auto* root=partition_by_exprs[0]->root();
    if (!root->is_slotref()) { return false;}
    auto* col_ref = down_cast<ColumnRef*>(root); 
    if (col_ref->slot_id()!=slot_id) {return false;}
} 
```

Comment (letian-jiang) [reply be/src/exprs/runtime_filter_bank.cpp:444]: should I also check `partition_by_exprs[1..]`? or checking `partition_by_exprs[0]` is enough?

Comment (letian-jiang) [reply be/src/exprs/runtime_filter_bank.cpp]: Update: traversed every expr to check only existent columns are referenced.

**Comment by LiShuMing** [comment]
LGTM

**Review by satanson — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Comment by letian-jiang** [comment]
@mergify backport branch-3.2

**Pull request merged**